src/isl/util.rs

```diff
@@ -698,33 +699,85 @@ pub enum TimestampPrecision {
     Millisecond,
     Microsecond,
     Nanosecond,
+    OtherFractionalSeconds(i64),
 }
 
 impl TimestampPrecision {
-    pub fn scale(timestamp_value: &Timestamp) -> i64 {
+    pub fn from_timestamp(timestamp_value: &Timestamp) -> TimestampPrecision {
+        use TimestampPrecision::*;
         let precision_value = timestamp_value.precision();
         match precision_value {
-            Precision::Year => -4,
-            Precision::Month => -3,
-            Precision::Day => -2,
-            Precision::HourAndMinute => -1,
-            Precision::Second => 0,
-            Precision::FractionalSeconds => timestamp_value.fractional_seconds_scale().unwrap(),
+            Precision::Year => Year,
+            Precision::Month => Month,
+            Precision::Day => Day,
+            Precision::HourAndMinute => Minute,
+            Precision::Second => Second,
+            Precision::FractionalSeconds => {
+                match timestamp_value.fractional_seconds_scale().unwrap() {
+                    3 => Millisecond,
+                    6 => Microsecond,
+                    9 => Nanosecond,
+                    scale => OtherFractionalSeconds(scale),
+                }
+            }
         }
     }
 }
 
-impl Into<i64> for TimestampPrecision {
-    fn into(self) -> i64 {
-        match self {
-            TimestampPrecision::Year => -4,
-            TimestampPrecision::Month => -3,
-            TimestampPrecision::Day => -2,
-            TimestampPrecision::Minute => -1,
-            TimestampPrecision::Second => 0,
-            TimestampPrecision::Millisecond => 3,
-            TimestampPrecision::Microsecond => 6,
-            TimestampPrecision::Nanosecond => 9,
-        }
+impl PartialOrd for TimestampPrecision {
+    fn partial_cmp(&self, other: &TimestampPrecision) -> Option<Ordering> {
```

Comment:
This arrangement works fine and is very explicit.

An alternative that's closer to your original implementation would be to map `self` and `other` to integers within this method and then use `.cmp` to produce the expected ordering. Something like:

```rust
let self_value = match self {
    Year => -4,
    Month => -3,
    // ...and so on
}
let other_value = match other {
    Year => -4,
    Month => -3,
    // ...and so on
}
self_value.cmp(other_value)
```